vali_objects/vali_dataclasses/perf_ledger.py

```diff
@@ -0,0 +1,100 @@
+
+    def get_or_create_latest_cp(self, now_ms: int):
+        if self.count_events() == 0 or self.cps[-1].accumulated_ms > self.target_cp_duration_ms:
+            new_cp = PerfCheckpoint(now_ms)
+            self.cps.append(new_cp)
+
+        return self.cps[-1]
+
+    def update_accumulated_time(self, cp: PerfCheckpoint, now_ms: int):
+        accumulated_time = now_ms - cp.last_update_time_ms
+        if accumulated_time < 0:
+            bt.logging.error(f"Negative accumulated time: {accumulated_time} for miner {self.miner_hotkey}."
+                             f" start_time_ms: {self.start_time_ms}, now_ms: {now_ms}")
+            accumulated_time = 0
+        cp.accumulated_ms += accumulated_time
+        cp.last_update_time_ms = now_ms
+        cp.n_updates += 1
+
+    def compute_return_between_ticks(self, current_portfolio_value: float, prev_portfolio_value: float):
+        return 1.0 + (current_portfolio_value - prev_portfolio_value) / prev_portfolio_value
+
+    def update_returns(self, current_cp: PerfCheckpoint, current_portfolio_value: float):
+        delta_return = self.compute_return_between_ticks(current_portfolio_value, current_cp.prev_portfolio_ret)
+        if current_portfolio_value >= current_cp.prev_portfolio_ret:
+            current_cp.gain_product *= delta_return
+        else:
+            current_cp.loss_product *= delta_return
+        current_cp.prev_portfolio_ret = current_portfolio_value
+
+    def purge_old_cps(self):
+        while self.get_total_ledger_duration_ms() > self.target_ledger_window_ms:
+            self.cps = self.cps[1:]  # Drop the first cp (oldest)
+
+    def update(self, current_portfolio_value):
+        now_ms = TimeUtil.now_in_millis()
+        current_cp = self.get_or_create_latest_cp(now_ms)
+        self.update_returns(current_cp, current_portfolio_value)
+        self.update_accumulated_time(current_cp, now_ms)
+        self.purge_old_cps()
+
+
+    def count_events(self):
+        # Return the number of events currently stored
+        return len(self.cps)
+
+    def get_product_of_gains(self):
+        ans = 1.0
+        for cp in self.cps:
+            ans *= cp.gain_product
+        return ans
+
+    def get_product_of_loss(self):
+        ans = 1.0
+        for cp in self.cps:
+            ans *= cp.loss_product
+        return ans
+
+    def get_total_ledger_duration_ms(self):
+        return sum(cp.accumulated_ms for cp in self.cps)
+
+def update_all_perf_ledgers(miner_hotkey_to_portfolio_value: dict[str, float]):
+    # Called every time MDD Checker runs. Currently means every 15-20 seconds.
+    miner_hotkey_to_perf_ledgers = ValiBkpUtils.safe_load_dict_from_disk('perf_ledgers.json', {})
+    for miner_hotkey, perf_ledger in miner_hotkey_to_perf_ledgers.items():
+        now_ms = TimeUtil.now_in_millis()
+        current_cp = perf_ledger.get_or_create_latest_cp(now_ms)
+        current_portfolio_value = miner_hotkey_to_portfolio_value.get(miner_hotkey, None)
```

Comment:
to be sure I've got this right - the checkpoints are going to look at the entirety of the portfolio value at every iteration? How much overhead would it bring to track a ledger for each tradepair